# Week 5, Tutorial 1: Clinical Data Literacy
**Author**: Sariana Ramoutar

The aim of this notebook is to become familiar with the structure and clinical meaning of the dataset before carrying out any cleaning or modelling. The focus is on understanding the available features, checking the data structure and identifying the target variable for future analysis.

# DATA / SAFETY NOTE
This is a de-identified US academic-hospital dataset; real identifiers (name, MRN, date of birth, address) have been removed.

## Import libraries

NumPy provides numerical functions, while pandas is used to load and explore the dataset.

In [23]:
import numpy as np    # Numerical helpers (e.g., NaN) are provided by NumPy.
import pandas as pd   # Pandas is used for data manipulation with DataFrames.

# Configure pandas to display more columns when printing DataFrames.
pd.set_option("display.max_columns", 60)
pd.set_option("display.width", 140)

print("Environment ready · pandas", pd.__version__)

Environment ready · pandas 2.2.2


## Data Loading and Structural Overview

The next step is loading the dataset then examining its overall shape (number of rows and columns), identifying all columns present, and understanding the data type of each column. This gives an overview of the dataset's structure.

In [24]:
from pathlib import Path
from google.colab import drive; drive.mount("/content/drive")
# Define the path to the dataset file, assuming it's in Google Drive.
DATA_PATH = Path('/content/drive/MyDrive/CariSurg Documents/yaleemmlc_admissionprediction_triage.csv')

# Check if the file exists and raise an error if not.
if not DATA_PATH.exists():
    raise FileNotFoundError(f"Could not find {DATA_PATH}. Place the CSV beside this notebook, or set DATA_PATH (Colab).")

# Load the CSV file into a pandas DataFrame.
# 'index_col=0' specifies that the first column should be used as the DataFrame index.
df = pd.read_csv(DATA_PATH, index_col=0)
print(f"Loaded {df.shape[0]:,} ED encounters  x  {df.shape[1]} columns")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Loaded 55,121 ED encounters  x  225 columns


In [25]:
# Display the first few rows of the DataFrame to inspect the data.
df.head()

,dep_name,esi,age,gender,ethnicity,race,lang,religion,maritalstatus,employstatus,insurance_status,disposition,arrivalmode,arrivalmonth,arrivalday,arrivalhour_bin,previousdispo,triage_vital_hr,triage_vital_sbp,triage_vital_dbp,triage_vital_rr,triage_vital_o2,triage_vital_o2_device,triage_vital_temp,triage_glucose,cc_abdominalcramping,cc_abdominaldistention,cc_abdominalpain,cc_abdominalpainpregnant,cc_abnormallab,...,cc_sorethroat,cc_stdcheck,cc_strokealert,cc_suicidal,cc_suture/stapleremoval,cc_swallowedforeignbody,cc_syncope,cc_tachycardia,cc_testiclepain,cc_thumbinjury,cc_tickremoval,cc_toeinjury,cc_toepain,cc_trauma,cc_unresponsive,cc_uri,cc_urinaryfrequency,cc_urinaryretention,cc_urinarytractinfection,cc_vaginalbleeding,cc_vaginaldischarge,cc_vaginalpain,cc_weakness,cc_wheezing,cc_withdrawal-alcohol,cc_woundcheck,cc_woundinfection,cc_woundre-evaluation,cc_wristinjury,cc_wristpain
7,A,4.0,87.0,Female,Hispanic or Latino,Other,Other,Pentecostal,Widowed,Retired,Medicare,Discharge,Car,March,Saturday,11-14,Admit,88.0,155.0,75.0,17.0,98.0,0.0,97.8,87.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
17,B,2.0,53.0,Male,Hispanic or Latino,Other,English,Catholic,Significant Other,Disabled,Medicare,Admit,Walk-in,September,Monday,11-14,Discharge,118.0,105.0,79.0,20.0,98.0,0.0,97.5,113.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
40,A,2.0,49.0,Female,Non-Hispanic,White or Caucasian,English,Catholic,Married,Full Time,Commercial,Discharge,ambulance,June,Tuesday,15-18,Admit,76.0,116.0,71.0,18.0,99.0,0.0,98.1,108.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
47,A,3.0,22.0,Female,Hispanic or Latino,White or Caucasian,English,Catholic,Single,Full Time,Commercial,Discharge,Car,April,Sunday,23-02,Discharge,106.0,103.0,63.0,16.0,97.0,0.0,98.2,85.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
60,A,2.0,62.0,Male,Non-Hispanic,White or Caucasian,English,Protestant,Divorced,Not Employed,Medicaid,Discharge,ambulance,July,Wednesday,23-02,Discharge,84.0,109.0,68.0,18.0,95.0,0.0,97.8,88.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### Observations

The dataset loaded successfully and contains approximately 55,000 emergency department encounters with 225 clinical features. Each row represents a single patient visit, while each column contains a recorded clinical variable or patient characteristic.

## Explore the dataset structure

The next step is to identify the different groups of variables within the dataset. Most columns contain structured demographic or clinical information, while a large proportion represent binary chief complaint indicators.

In [26]:
# Print the total number of columns in the DataFrame.
print("Total columns:", df.shape[1], "\n")

# Separate columns into 'structured' (non-chief complaint) and 'cc_flags' (chief complaint).
structured = [c for c in df.columns if not c.startswith("cc_")]
cc_flags   = [c for c in df.columns if c.startswith("cc_")]

# Print the structured columns and the count of chief complaint flags.
print(f"Structured columns ({len(structured)}):")
for c in structured:
    print("   ", c)
print(f"\nChief-complaint flags (cc_*): {len(cc_flags)}  ->  grouped by body system in section 5")

Total columns: 225 

Structured columns (25):
    dep_name
    esi
    age
    gender
    ethnicity
    race
    lang
    religion
    maritalstatus
    employstatus
    insurance_status
    disposition
    arrivalmode
    arrivalmonth
    arrivalday
    arrivalhour_bin
    previousdispo
    triage_vital_hr
    triage_vital_sbp
    triage_vital_dbp
    triage_vital_rr
    triage_vital_o2
    triage_vital_o2_device
    triage_vital_temp
    triage_glucose

Chief-complaint flags (cc_*): 200  ->  grouped by body system in section 5


### Observations

The dataset contains a relatively small number of structured variables alongside approximately 200 chief complaint indicators. This suggests the dataset is wide, with patient presentations represented primarily through binary complaint flags.

## Examine data types

Understanding the data type of each variable helps determine how it should be analysed. Numerical variables can be summarised statistically, whereas categorical variables are generally explored using frequencies.

In [27]:
# Display the count of each data type present in the DataFrame.
print("Column types at a glance:")
print(df.dtypes.value_counts(), "\n")     # how many numeric vs text columns

# Provide a concise summary of the structured columns, including non-null counts and data types.
df[structured].info()

# Display the first 3 rows for a selection of key columns for a quick overview.
df[["esi", "age", "gender", "race", "triage_vital_hr", "triage_vital_sbp", "disposition"]].head(3)

Column types at a glance:
float64    210
object      15
Name: count, dtype: int64 

<class 'pandas.core.frame.DataFrame'>
Index: 55121 entries, 7 to 433332
Data columns (total 25 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   dep_name                55121 non-null  object 
 1   esi                     55121 non-null  float64
 2   age                     55121 non-null  float64
 3   gender                  55121 non-null  object 
 4   ethnicity               55121 non-null  object 
 5   race                    55121 non-null  object 
 6   lang                    55121 non-null  object 
 7   religion                55121 non-null  object 
 8   maritalstatus           55121 non-null  object 
 9   employstatus            55121 non-null  object 
 10  insurance_status        55121 non-null  object 
 11  disposition             55121 non-null  object 
 12  arrivalmode             55121 non-null  object 
 13  arrivalmont

,esi,age,gender,race,triage_vital_hr,triage_vital_sbp,disposition
7,4.0,87.0,Female,Other,88.0,155.0,Discharge
17,2.0,53.0,Male,Other,118.0,105.0,Admit
40,2.0,49.0,Female,White or Caucasian,76.0,116.0,Discharge


### Observations

The dataset contains both numerical and categorical variables. The structured clinical fields appear to have appropriate data types, providing a suitable foundation for further profiling and quality assessment.

## Patient Demographics and Outcomes

This section explores the demographic information recorded for each patient together with the outcome variables. These variables describe who the patient is and what happened during the emergency department visit.
-   **`age`**: Patient's age in years. This dataset specifically contains `adults only` (aged 18 or above), which is a crucial demographic detail.
-   **`gender`**: Recorded as Male/Female. This feature is clinically significant and also an important axis for assessing fairness.
-   **`ethnicity`, `race`**: Self-identified categories that are `fairness-sensitive` and require careful consideration to prevent bias in modelling.
-   **`disposition`**: Describes the outcome of the visit (e.g., Admit/Discharge). This is an `outcome` determined hours after triage and should not be used as an input for a triage prediction model.
-   **`esi` (Triage_Level)**: The assigned triage acuity level. This is the `target variable` for our predictive model.

In [28]:
# For numerical 'age', display descriptive statistics.
print("AGE (years):")
print(df["age"].describe().round(1), "\n")

# For categorical columns, count the occurrences of each unique value, including missing ones.
for col in ["gender", "race", "ethnicity", "disposition"]:
    print(f"{col}:")
    print(df[col].value_counts(dropna=False), "\n")

AGE (years):
count    55121.0
mean        55.3
std         19.5
min         18.0
25%         40.0
50%         55.0
75%         70.0
max        107.0
Name: age, dtype: float64 

gender:
gender
Female    31744
Male      23377
Name: count, dtype: int64 

race:
race
White or Caucasian                           29435
Black or African American                    15963
Other                                         9016
Patient Refused                                370
Asian                                          175
Unknown                                         76
American Indian or Alaska Native                66
Native Hawaiian or Other Pacific Islander       20
Name: count, dtype: int64 

ethnicity:
ethnicity
Non-Hispanic          45142
Hispanic or Latino     9888
Patient Refused          56
Unknown                  35
Name: count, dtype: int64 

disposition:
disposition
Discharge    34565
Admit        20556
Name: count, dtype: int64 



### Observations

The dataset contains adult patients only. Age is recorded as a numerical variable, while gender, race and ethnicity are categorical. These variables may be clinically informative but should also be considered carefully when assessing fairness and potential bias in later modelling.

## Vital signs

Vital signs recorded during triage provide an objective assessment of the patient's physiological condition on arrival. These measurements are expected to contribute strongly to emergency department triage decisions.

| Column | Vital | Normal adult Range | Clinical Significance |
|---|---|---|---|
| `triage_vital_hr` | Heart rate (bpm) | 60–100 | > 100 indicates tachycardia, a potential sign of stress or shock. |
| `triage_vital_sbp` | Systolic BP (mmHg) | 90–140 | < 90 suggests possible shock. |
| `triage_vital_dbp` | Diastolic BP (mmHg) | 60–90 | Interpreted alongside SBP for overall blood pressure assessment. |
| `triage_vital_rr` | Respiratory rate (/min) | 12–20 | One of the earliest indicators of clinical deterioration. |
| `triage_vital_o2` | Oxygen saturation (%) | 95–100 | < 92 is concerning, indicating hypoxia. |
| `triage_vital_temp` | Temperature (**°F**) | 97.0–99.5 | Crucially, this is in `Fahrenheit` (not Celsius); fever is typically ≥ 100.4 °F (38 °C). |
| `triage_glucose` | Point-of-care glucose (mg/dL) | 70–140 | Both extremely high and low values are clinically significant. |

Additionally, `triage_vital_o2_device` (coded 0/1) indicates whether supplemental oxygen was in use during SpO₂ measurement, which impacts the interpretation of the oxygen saturation value.

In [29]:
# Define normal adult ranges for key vital signs, noting temperature is in Fahrenheit.
NORMAL_RANGES = {
    "triage_vital_hr":   (60, 100, "bpm"),
    "triage_vital_sbp":  (90, 140, "mmHg"),
    "triage_vital_dbp":  (60,  90, "mmHg"),
    "triage_vital_rr":   (12,  20, "/min"),
    "triage_vital_o2":   (95, 100, "%"),
    "triage_vital_temp": (97.0, 99.5, "F"),   # NOT Celsius
    "triage_glucose":    (70, 140, "mg/dL"),
}
VITALS = list(NORMAL_RANGES)

# Display descriptive statistics for the vital signs.
print(df[VITALS].describe().round(1))

# Count the occurrences of 'triage_vital_o2_device' to see how many patients were on supplemental oxygen.
print("\nOn a supplemental-O2 device?  (0 = room air, 1 = device)")
print(df["triage_vital_o2_device"].value_counts(dropna=False))

       triage_vital_hr  triage_vital_sbp  triage_vital_dbp  triage_vital_rr  triage_vital_o2  triage_vital_temp  triage_glucose
count          55121.0           55121.0           55121.0          55121.0          55121.0            55121.0         55121.0
mean              86.4             133.7              79.5             17.8             97.0               98.1           130.1
std               17.0              22.6              14.6              2.1              2.1                0.8            73.7
min               32.0              53.0              27.0              8.0             62.0               91.5            16.0
25%               74.0             118.0              70.0             16.0             96.0               97.7            93.0
50%               85.0             132.0              79.0             18.0             98.0               98.0           107.0
75%               97.0             147.0              89.0             18.0             98.0            

### Observations

The recorded vital signs appear clinically plausible and cover the major physiological measurements collected during triage. Temperature is recorded in Fahrenheit rather than Celsius, which should be considered during interpretation and any future preprocessing.

## Chief complaint indicators

Chief complaints are recorded as binary variables indicating whether a particular complaint was documented during triage. Rather than examining each complaint individually, grouping them into broader body systems provides a more meaningful overview of patient presentations.

In [30]:
cc_flags = [c for c in df.columns if c.startswith("cc_")]

# Define categories for chief complaints using keywords. Specific keywords are prioritised.
CC_CATEGORIES = [
    ("Cardiovascular",             ["chestpain","chesttightness","palpitation","irregularheart","rapidheart","tachycard","hypertens","hypotens","cardiacarrest"]),
    ("Genitourinary/Renal",        ["dysuria","hematuria","urinary","flank","testicle","maleguproblem","femaleguproblem","groin","pelvic","vaginal","breastpain","std"]),
    ("Respiratory",                ["breathing","shortnessofbreath","dyspnea","cough","wheez","respiratory","asthma","hemoptysis","coldlike","nasalcongest","influenza","sinus","sorethroat","uri"]),
    ("Neurological",               ["headache","migraine","dizz","syncope","seizure","stroke","alteredmental","confusion","numbness","lossofconsciousness","unresponsive","lethargy","neurologic","hallucinat","extremityweakness"]),
    ("Gastrointestinal",           ["abdominal","epigastric","nausea","emesis","vomit","diarrhea","constipation","gibleeding","giproblem","rectal","swallowedforeign","ingestion","dehydration"]),
    ("Psych/Behavioral/Substance", ["anxiety","depress","agitation","suicid","homicid","psychot","psychiatric","panic","alcohol","drugproblem","drug/alcohol","addiction","detox","overdose","withdrawal","poison"]),
    ("ENT/Eye/Dental",             ["earpain","earproblem","otalgia","eye","conjunctiv","foreignbodyineye","blurredvision","dental","jaw","epistaxis","facial","oralswelling"]),
    ("Skin/Soft-tissue",           ["rash","cellulitis","abscess","cyst","skin","wound","mass","allergicreaction","edema","bruising"]),
    ("Endocrine/Metabolic/Heme",   ["bloodsugar","glycem","glucose","sicklecell"]),
    ("Constitutional/General",     ["fever","chills","fatigue","bodyache","generalizedbody","weakness"]),
    ("Trauma/Injury/MSK",          ["injury","fall","trauma","laceration","fracture","back","neck","joint","rib","assault","burn","animalbite","insectbite","pain","swelling"]),
]

# Function to categorise a chief complaint flag into its body system.
def categorize(flag):
    stem = flag[3:]                     # Remove the "cc_" prefix.
    if "crash" in stem:                 # Handle specific cases like 'crash' before broader terms.
        return "Trauma/Injury/MSK"
    for category, keywords in CC_CATEGORIES:
        if any(k in stem for k in keywords):
            return category
    return "Other/Procedural/Admin"

# Group chief complaint flags by their assigned categories.
from collections import defaultdict
groups = defaultdict(list)
for flag in cc_flags:
    groups[categorize(flag)].append(flag)

# Print the count of flags in each category, ensuring all flags are categorised.
order = [name for name, _ in CC_CATEGORIES] + ["Other/Procedural/Admin"]
for name in order:
    print(f"{name:28s} {len(groups[name]):3d} flags")
print("-" * 40)
print(f"{"TOTAL placed":28s} {sum(len(v) for v in groups.values()):3d} of {len(cc_flags)}")

Cardiovascular                 9 flags
Genitourinary/Renal           17 flags
Respiratory                   15 flags
Neurological                  21 flags
Gastrointestinal              16 flags
Psych/Behavioral/Substance    18 flags
ENT/Eye/Dental                18 flags
Skin/Soft-tissue              14 flags
Endocrine/Metabolic/Heme       5 flags
Constitutional/General         8 flags
Trauma/Injury/MSK             49 flags
Other/Procedural/Admin        10 flags
----------------------------------------
TOTAL placed                 200 of 200


### Observations

Chief complaint indicators account for most of the variables in the dataset. Grouping them by body system provides a clearer summary of patient presentations and highlights which clinical categories occur most frequently.

In [31]:
# (1) Compute the total number of times complaint flags fired within each body system.
print("Complaint volume by body system (flag-hits across all encounters):")
system_totals = {name: int(df[flags].sum().sum()) for name, flags in groups.items()}
# Print system totals, sorted from most to least common.
for name in sorted(system_totals, key=system_totals.get, reverse=True):
    print(f"   {name:28s} {system_totals[name]:>7,}")

# (2) Identify and print the top 5 most common individual complaints.
per_flag = df[cc_flags].sum().sort_values(ascending=False)
print("\nTop 5 individual complaints:")
print(per_flag.head(5))
# Count and print any chief complaint flags that never appeared in the dataset.
print("\nComplaint flags that never fire in this sample:", int((per_flag == 0).sum()))

Complaint volume by body system (flag-hits across all encounters):
   Trauma/Injury/MSK             13,094
   Gastrointestinal              10,542
   Respiratory                    6,693
   Other/Procedural/Admin         6,551
   Neurological                   5,957
   Cardiovascular                 4,912
   Psych/Behavioral/Substance     4,189
   Genitourinary/Renal            3,418
   Constitutional/General         3,032
   Skin/Soft-tissue               2,464
   ENT/Eye/Dental                 1,571
   Endocrine/Metabolic/Heme         875

Top 5 individual complaints:
cc_abdominalpain        6717.0
cc_other                4491.0
cc_chestpain            3712.0
cc_shortnessofbreath    3098.0
cc_backpain             1997.0
dtype: float64

Complaint flags that never fire in this sample: 1


### Observations

The analysis of chief complaints clearly shows that 'Trauma/Injury/MSK', 'Gastrointestinal', and 'Respiratory' are the body systems with the highest reported complaint volumes. Individually, 'abdominalpain', 'other', 'chestpain', and 'shortnessofbreath' are the most frequent complaints. It's also notable that one complaint flag never fired in this dataset, indicating it was not recorded for any patient within this sample.

## Emergency Severity Index (ESI)

The Emergency Severity Index (ESI) represents the triage category assigned to each patient at presentation. As this decision is made during the initial assessment, it is the most appropriate target variable for future predictive modelling.

In [32]:
# Calculate the value counts for ESI and convert to percentages.
counts = df["esi"].value_counts(dropna=False).sort_index()
pct    = (counts / len(df) * 100).round(1)
print("ESI (Triage_Level) distribution:")
# Print the distribution of ESI levels, including counts and percentages.
for level in counts.index:
    print(f"   ESI {level:>3}   {counts[level]:>7,}   ({pct[level]:>4}%)")
# Check for and report any missing ESI labels.
print("\nMissing labels:", int(df["esi"].isna().sum()))

ESI (Triage_Level) distribution:
   ESI 1.0        77   ( 0.1%)
   ESI 2.0    17,924   (32.5%)
   ESI 3.0    27,010   (49.0%)
   ESI 4.0     8,896   (16.1%)
   ESI 5.0     1,214   ( 2.2%)

Missing labels: 0


### Observations

The ESI variable contains no missing values and therefore provides a complete target for supervised learning. In contrast, disposition represents a later clinical outcome and should not be used as an input feature because it would introduce data leakage.

## ESI Severity Levels

The ESI system classifies patients into five levels according to the urgency of their condition. Lower scores represent higher clinical acuity, while higher scores represent less urgent presentations.

In [33]:
# Define a mapping from ESI numerical levels to their descriptive names.
labels = {1: "Resuscitation", 2: "Emergent", 3: "Urgent", 4: "Less urgent", 5: "Non-urgent"}
# Get the counts of each ESI level and sort by index.
counts = df["esi"].value_counts().sort_index()
print("ESI  ·  name           ·  encounters")
# Print each ESI level with its name and the number of encounters.
for level, n in counts.items():
    print(f"   {int(level)}  ·  {labels[int(level)]:<13}  ·  {n:>7,}")

ESI  ·  name           ·  encounters
   1  ·  Resuscitation  ·       77
   2  ·  Emergent       ·   17,924
   3  ·  Urgent         ·   27,010
   4  ·  Less urgent    ·    8,896
   5  ·  Non-urgent     ·    1,214


### Observations

Most patients fall within the middle acuity levels, while the highest priority cases are comparatively uncommon. This distribution is consistent with typical emergency department activity.